File: java/src/main/java/com/tormentnexus/goose/ProviderInterface.java
package com.tormentnexus.goose;

// Re-implementation of Goose's Abstract LLM Provider Interface

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

public interface ProviderInterface {

    enum Role {
        USER, ASSISTANT, SYSTEM
    }

    class Message {
        public Role role;
        public String content;

        public Message(Role role, String content) {
            this.role = role;
            this.content = content;
        }
    }

    class ProviderResponse {
        public String text;
        public Integer promptTokens;
        public Integer completionTokens;

        public ProviderResponse(String text, Integer promptTokens, Integer completionTokens) {
            this.text = text;
            this.promptTokens = promptTokens;
            this.completionTokens = completionTokens;
        }
    }

    interface LLMProvider {
        String getName();
        CompletableFuture<ProviderResponse> chat(List<Message> messages);
        CompletableFuture<ProviderResponse> streamChat(List<Message> messages, Consumer<String> onChunk);
    }
}

File: rust/src/goose/provider_interface.rs
// Re-implementation of Goose's Abstract LLM Provider Interface

use async_trait::async_trait;

pub enum Role {
    User,
    Assistant,
    System,
}

pub struct Message {
    pub role: Role,
    pub content: String,
}

pub struct ProviderResponse {
    pub text: String,
    pub prompt_tokens: Option<u32>,
    pub completion_tokens: Option<u32>,
}

#[async_trait]
pub trait LLMProvider {
    fn name(&self) -> &str;
    async fn chat(&self, messages: &[Message]) -> Result<ProviderResponse, String>;
    async fn stream_chat<F>(&self, messages: &[Message], on_chunk: F) -> Result<ProviderResponse, String>
    where
        F: FnMut(&str) + Send + Sync;
}
